hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieHFileReader.java

```diff
@@ -329,39 +410,85 @@ public R next() {
+    //       to avoid seeking back and forth
+    Collections.sort(keyPrefixes);
+    List<Pair<byte[], byte[]>> keyRecordsBytes = new ArrayList<>(keyPrefixes.size());
+    for (String keyPrefix : keyPrefixes) {
+      KeyValue kv = new KeyValue(keyPrefix.getBytes(), null, null, null);
+      int val = hFileScanner.seekTo(kv);
+      // what does seekTo() does:
+      // eg entries in file. [key01, key02, key03, key04,..., key20]
+      // when keyPrefix is "key", seekTo will return -1 and place the cursor just before key01. getCel() will return key01 entry
+      // when keyPrefix is ""key03", seekTo will return 0 and place the cursor just before key01. getCell() will return key03 entry
+      // when keyPrefix is ""key1", seekTo will return 1 and place the cursor just before key10(i.e. key09). call next() and then call getCell() to see key10 entry
+      // when keyPrefix is "key99", seekTo will return 1 and place the cursor just before last entry, ie. key04. getCell() will return key04 entry.
+
+      if (val == 1) { // move to next entry if return value is 1
+        if (!hFileScanner.next()) {
+          // we have reached the end of file. we can skip proceeding further
+          break;
+        }
       }
+      do {
+        Cell c = hFileScanner.getCell();
+        byte[] keyBytes = Arrays.copyOfRange(c.getRowArray(), c.getRowOffset(), c.getRowOffset() + c.getRowLength());
+        String key = new String(keyBytes);
+        // Check whether we're still reading records corresponding to the key-prefix
+        if (!key.startsWith(keyPrefix)) {
+          break;
+        }
+
+        // Extract the byte value before releasing the lock since we cannot hold on to the returned cell afterwards
+        byte[] valueBytes = Arrays.copyOfRange(c.getValueArray(), c.getValueOffset(), c.getValueOffset() + c.getValueLength());
+        keyRecordsBytes.add(Pair.newPair(keyBytes, valueBytes));
+      } while (hFileScanner.next());
     }
-    return false;
+
+    // Use  tree map so that entries are in sorted in the map being returned.
+    Map<String, R> values = new TreeMap<String, R>();
+    for (Pair<byte[], byte[]> kv : keyRecordsBytes) {
+      R record = deserialize(kv.getFirst(), kv.getSecond(), readerSchema, readerSchema, keyFieldSchema);
+      values.put(new String(kv.getFirst()), record);
+    }
+    return values;
   }
 
   @Override
-  public Option getRecordByKey(String key, Schema readerSchema) throws IOException {
+  public Option<R> getRecordByKey(String key, Schema readerSchema, HFileScanner hFileScanner, Option<Schema.Field> keyFieldSchema) throws IOException {
     byte[] value = null;
-    final Option<Schema.Field> keyFieldSchema = Option.ofNullable(readerSchema.getField(KEY_FIELD_NAME));
-    ValidationUtils.checkState(keyFieldSchema != null);
     KeyValue kv = new KeyValue(key.getBytes(), null, null, null);
 
-    synchronized (this) {
-      if (keyScanner == null) {
```

Comment:
removed the class instance HFileScanner so that concurrent readers can use their own HfileScanner. and hence no synchronization required.